tests/admin_changelist/tests.py

```diff
@@ -937,6 +939,89 @@ def check_results_order(ascending=False):
         OrderedObjectAdmin.ordering = ['id', 'bool']
         check_results_order(ascending=True)
 
+    @isolate_apps('admin_changelist')
+    def test_total_ordering_optimization(self):
+        class Related(models.Model):
+            unique_field = models.BooleanField(unique=True)
+
+            class Meta:
+                ordering = ('unique_field',)
+
+        class Model(models.Model):
+            unique_field = models.BooleanField(unique=True)
+            unique_nullable_field = models.BooleanField(unique=True, null=True)
+            related = models.ForeignKey(Related, models.CASCADE)
+            other_related = models.ForeignKey(Related, models.CASCADE)
+            related_unique = models.OneToOneField(Related, models.CASCADE)
+            field = models.BooleanField()
+            other_field = models.BooleanField()
+            null_field = models.BooleanField(null=True)
+
+            class Meta:
+                unique_together = {
+                    ('field', 'other_field'),
+                    ('field', 'null_field'),
+                    ('related', 'other_related_id')
+                }
+
+        queryset = Model.objects.none()
+
+        class ModelAdmin(admin.ModelAdmin):
+            def get_queryset(self, request):
+                return queryset
+
+        request = self._mocked_authenticated_request('/', self.superuser)
+        site = admin.AdminSite(name='admin')
+        model_admin = ModelAdmin(Model, site)
+        change_list = model_admin.get_changelist_instance(request)
+        self.assertEqual(change_list._get_deterministic_ordering([]), ['-pk'])
```

Comment:
Did you consider a loop with subTest for all these assertions?

Reply:
TBH, didn't think about it. 

I quite liked the presentation (with the comments) when first reading it through: helped me reason about it.

Reply:
Sure, we sometimes use comments with that pattern;
```
tests = (
    # Unique non-nullable field.
    (['unique_field']), ['unique_field']),
    ...
    # ....
    ...
)
```

Reply:
OK. That makes sense. Let me have a go at re-writing that.

Reply:
https://github.com/django/django/pull/10692/commits/deadb225ccb703b24ee685e079930541149e2370 (which is a `fixup`)